fix(css-shorthands): expanding the shorthand properties in the styles

mnajdova opened the pull request:
## Fix
This PR aims to fix https://github.com/stardust-ui/react/issues/237 by expanding the css shorthands of the theme's components styles and the styles prop provided to the component while they are merged.

I added the css-shorthand-expand library and provided service for expanding the styles. There are still some points that should be addressed:
### 1. The library provides extending for the following shorthands: 

```
  font
  padding
  margin
  border
  borderWidth
  borderStyle
  borderColor
  borderTop
  borderRight
  borderBottom
  borderLeft
  borderRadius
  background
  outline
```

This means that we still don't have any mechanism for expanding the other shorthands like:
 
```
  animation
  column-rule
  columns flex
  flex-flow grid
  grid-area
  grid-column
  grid-row
  grid-template
  list-style
  offset
  overflow
  place-content
  place-items
  place-self
  text-decoration
  transition
```

### 2. It would have been nice if we would be able to measure the performance of the components and conclude whether this changes would impact them.

I need input whether this is worth exploring further, or whether maybe we should just try to use the library when defining the styles for the themes or applying the styles prop on the components.

# NOTE
The css-shorthand-expand library was replaced with a better alternative shortcss. There is one known bug in the first library that is preventing us of using it. I liked the second library better because of more cleaner expanding of the props.

Commit:
-added css-shorthand-expand dependency and incorporated it in the mergeTheme logic

Commit:
-introduced the logic for expanding the css shorthands in a fela plugin

kuzhelov commented on src/lib/felaExpandCssShorthandsPlugin.ts:
```diff
@@ -0,0 +1,47 @@
+    Object.keys(styles).forEach(cssPropertyName => {
+      const cssPropertyValue = styles[cssPropertyName]
+
+      if (typeof cssPropertyValue === 'object') {
```
to be sure, could you, please, suggest what is the case for it? are we targeting pseudo element styles with this check? In that case we might omit it, given that there is this specific names filter that comes next:

```js
_.includes(shorthands, cssPropertyName)
```

kuzhelov commented on src/lib/felaExpandCssShorthandsPlugin.ts:
```diff
@@ -0,0 +1,47 @@
+  const expandCssShorthands = styles => {
+    const processedStyles = {}
+
+    Object.keys(styles).forEach(cssPropertyName => {
```
nit: if `lodash` is imported and, what is more appealing, is used below, we might consider to use _.keys :) in either case, please, feel free to decide by your own

mnajdova replied:
The code is refactored.

kuzhelov commented on src/lib/felaExpandCssShorthandsPlugin.ts:
```diff
@@ -0,0 +1,47 @@
+
+export default () => {
+  const expandCssShorthands = styles => {
+    const processedStyles = {}
```
this logic is overall a good candidate for `reduce`:

```js
return _.keys(styles).reduce((acc, cssPropertyName) => {
  const cssPropertyValue = styles[cssPropertyName]

   if (_.includes(shorthands, cssPropertyName)) {
        const expandedProps = expand(_.kebabCase(cssPropertyName), String(cssPropertyValue))
        Object.keys(expandedProps).forEach(key => {
          processedStyles[_.camelCase(key)] = expandedProps[key]
        })

        return { ...acc, ...expandedProps }
      }

  return  { ...acc,  { cssPropertyName: cssPropertyValue }}
}, {})

mnajdova replied:
I refactored the whole code and removed the shorthands array.

kuzhelov commented on src/lib/felaExpandCssShorthandsPlugin.ts:
```diff
@@ -0,0 +1,47 @@
+  return expandCssShorthands
+}
+
+const shorthands = [
```
could you, please, suggest which cases we are aim to prevent with this explicit list?

mnajdova replied:
I deleted it

kuzhelov commented on src/lib/felaRenderer.tsx:
```diff
@@ -9,6 +10,8 @@ import { IRenderer } from '../../types/theme'
 
 const createRendererConfig = (options: any = {}) => ({
   plugins: [
+    felaExpandCssShorthandsPlugin(),
```
shouldn't it come after CSS is sanitized?

Commit:
-refactored the expand css shorthand fela plugin
-moved it after the sanitize plugin

Commit:
-added logic for expanding the shorthand css props in the nested pseudo classes too
-added test for the new fela plugin

Commit:
Merge branch 'master' into bug/css-shorthand

# Conflicts:
#	package.json

Commit:
-replaced css-shorthand-props with shortcss (because if known bug and not released newer version after it is fixed)

Commit:
-fixed checking for a shorthand prop
-added tests for covering the borderColor prop (as an example for ensuring two-words properties work)

Commit:
-better description of the tests
